### ISSUE TITLE
adding user ext to style configurator does not take effect for files unless each file is closed and reopened

### ISSUE THREAD

**Issue by codee-weber** [notepad-plus-plus/notepad-plus-plus#11830]
### Steps to Reproduce the Issue

1. Click Settings > Style Configurator > Language
2. Select a language, e.g. HTML
3. Type a new extension in "User Ext" e.g. "hta" (if there is already an extension in "User Ext", append the new one to the end separated by a space character)
4. Click [Save and Close]

### Expected Behavior

Files with the new extension (e.g. "hta") should now appear formatted per the associated language.

### Actual Behavior
File formatting does not change. 
Closing and re-opening Notepad++ does not help. 
Through trial and error, we find that **each** file must be _individually_ closed and re-opened, in order for the formatting to take effect.

### Debug Information

Notepad++ v8.3.3   (32-bit)
Build time : Mar 13 2022 - 17:11:10
Path : C:\Program Files (x86)\Notepad++\notepad++.exe
Command Line : 
Admin mode : ON
Local Conf mode : OFF
Cloud Config : OFF
OS Name : Windows 10 Enterprise (64-bit) 
OS Version : 2009
OS Build : 19044.1766
Current ANSI codepage : 1252
Plugins : DSpellCheck.dll mimeTools.dll NppConverter.dll NppExec.dll NppExport.dll 

I'm not sure if this is a "bug" per se, or just a "missing feature", but it is damn annoying, and if you google "Associate file extensions to language in Notepad++" you will see that a lot of other people have had the same confusion. They ask how to do it, and people tell them to go to the Style Configurator, find the language, and add the extension in "User Ext" and it doesn't work. Many users end up frustrated and never figure out they need to close and re-open every file individually (which is a big pain if you have 40 files open in the editor). 
Hope this helps!

**Comment by TobiasKnauss** [comment]
+1
had the same issue today. Even `File -> Reload from disk` does not help, the new style is not applied.
NPP 8.9.1 x64

In `Steps to Reproduce the Issue`, the OP forgot to mention that you need to open files with the relevant extension first.

**Comment by pryrt** [comment]
Notepad++ does the extension-based decision making when a file is opened.  Otherwise, if you manually set the language of a given file, at some random point Notepad++ might change it back, because it decided it knew better than you what language a given file should be.  

If you've only got a few of that file extension open, then it's easiest to close each file and re-open them (easy enough with `Ctrl+W` then `Ctrl+Shift+T`).  

But if, like the original poster, you already have 40 files with that extension open, then I can see why that would be annoying.  (Though I cannot see why you'd open all 40 files, _then_ decide, "no, I want these all to be some specific syntax highlighting").  Under such circumstances, I could see that a button on the Style Configurator that executes "apply this Language to all open files matching these extensions" might be reasonable... though I don't have a good name to be given to that button -- **Apply Extensions** is the best I can come up with, but I don't it very much.

That said, it appears that over the last four or so years, there's been a total of 2 people that have been interested in this feature.  The original post indicated that lots of people were frustrated, but I cannot recall ever seeing that frustration expressed here or in the Community Forum.  (Obviously, there would be a lot of posts in various places asking about how to assign the extension -- but in every one I found, using the google search the OP suggested, someone in the answers or comments pointed out you have to re-open, and I never saw anyone express frustration at it).  In your experience, is it really that common to have more than a couple of the same file extension open _before_ assigning that extension to the desired Language?  (Without evidence that a lot of people want it, no developer or volunteer is likely to work on this issue, as can be seen by 4 years of no reply.)

--
_edit_: fixed shortcut from `Ctrl+T` to `Ctrl+Shift+T`

**Comment by TobiasKnauss** [comment]
@pryrt I agree that the scenario of tens of open files might be very rare. I had 5 such files open.

The ideas that came to my mind:
a) Add an option in the settings that a file reload (from the FILE menu) would be handled like a manual close and reopen.
b) Add a note below the `user ext` textbox, saying: "The changes on file extensions are NOT automatically applied to loaded files. They must be closed and reopened manually." or something similar. This would already help avoiding the confusion.
c) Add a parameter in the session.xml, that remembers if a language was chosen manually. If the language was chosen automatically, apply the new extensions after the settings were changed.
d) Add the button you suggested.
e) After looking at the LANGUAGE menu, I don't see any option to revert to the automatically detected language. This should be added in any case. Having a keyboard shortcut for it, you could - in combination with Ctrl+PgUp - quickly iterate over all loaded tabs and reload the default language.
f) any combination of those.

**Comment by pryrt** [comment]
> in combination with Ctrl+PgUp - quickly iterate over all loaded tabs and reload the default language.

Record Macro: ~~`Ctrl+W` , `Ctrl+Shift+T` , `Ctrl+PgUp`~~, assign to your desired shortcut, and use that shortcut N times to reload N files (or **Macro > Run a Macro Multiple Times**).  Requires no changes to Notepad++.

--
_edit:_ Fix `Ctrl+T` to `Ctrl+Shift+T` in this reply and previous

_edit 2:_ Two problems with the recorded macro: 
1. `Ctrl+Shift+T` (IDM_FILE_RESTORELASTCLOSEDFILE) is apparently not macro-recordable, though it is playable.  So manually editing the macro to use wParam="41021" will fix that.
2. With that sequence, since hte restore-last-closed-file opens the file at the end, the `Ctrl+PgUp` can just end up swapping between two files.  Using `Ctrl+PgDn` instead will avoid that issue.

Thus, the macro that works for me:
```
        <Macro name="ReopenDn" Ctrl="yes" Alt="yes" Shift="yes" Key="84">
            <Action type="2" message="0" wParam="41003" lParam="0" sParam="" />
            <Action type="2" message="0" wParam="41021" lParam="0" sParam="" />
            <Action type="2" message="0" wParam="44095" lParam="0" sParam="" />
        </Macro>
```
Paste that into `%AppData%\Notepad++\shortcuts.xml` at the end of the `<Macros>` section, save, exit Notepad++, and run Notepad++, and now `Ctrl+Alt+Shift+T` will close and reopen the current file, and move to the next one to the right.  So if you run that N times, it will reopen all N files.

**Comment by TobiasKnauss** [comment]
Nice solution!
The only drawback: you loose the current scroll position in the file. Due to this, it would be great to have my idea e) implemented, which then would solve the OP's issue with a script that runs `apply default language` + `move to next file`.

However, the hint in the settings should still be added. And this macro is not that easy to create for macro beginners.

**Comment by pryrt** [comment]
> However, the hint in the settings should still be added. 

If the developer things such a hint as a good idea, I won't fight it.  But, IMO, you shouldn't go looking for documentation hints in the user interface itself; if there's something a user doesn't understand, the User Manual seems a better place to document it.  (And I'm about to go check, and if it's not already mentioned in the User Manual, I will add it in the next few minutes.)

> And this macro is not that easy to create for macro beginners.

I thought my edit2-instructions could be followed, even by a beginner.  If you disagree, then I guess I have more faith in the ability of even beginners to follow instructions.

And I wasn't saying that the macro is the end-all-and-be-all solution; but for people who want something more than what currently exists, it might be useful -- especially since it might be years (or never) before one or more of your suggestions are implemented.

**Comment by TobiasKnauss** [comment]
I don't argue against user's ability to copy a text and paste it as a macro. The challenge would be finding this issue in the first step. So a user might rather try to create its own.
I was not offended by the first version of your last comment, you could have kept it. It should be fine to express your opinion in that way without hurting someone...

However, I do agree that the hint should be in the manual. Although I'm a software developer myself, I have to admit that I often forget to look in the manual, because I try to create UIs of my software as intuitive as possible so that they often don't need a manual. Users usually don't like to RTFM... ;-)

Yes, IF someone finds time for it...  therefore, like I said: nice solution! It does the job!
